components/blitz/src/omero/gateway/facility/DataManagerFacility.java

```diff
@@ -497,4 +572,242 @@ public void addImagesToDataset(SecurityContext ctx,
     }
 
+    /**
+     * Creates the {@link DatasetData} on the server and attaches it
+     * to the {@link ProjectData} (if not <code>null</code>) (if the 
+     * project doesn't exist on the server yet, it will be created, too)
+     * @param ctx The {@link SecurityContext}
+     * @param dataset The {@link DatasetData}
+     * @param project The {@link ProjectData}
+     * @return The {@link DatasetData}
+     * @throws DSOutOfServiceException
+     *             If the connection is broken, or not logged in
+     * @throws DSAccessException
+     *             If an error occurred while trying to retrieve data from OMERO
+     *             service.
+     */
+    public DatasetData createDataset(SecurityContext ctx, DatasetData dataset,
+            ProjectData project) throws DSOutOfServiceException,
+            DSAccessException {
+        if (project != null) {
+            ProjectDatasetLink link = new ProjectDatasetLinkI();
+            link = new ProjectDatasetLinkI();
+            link.setChild(dataset.asDataset());
+            link.setParent(project.asProject());
+            link = (ProjectDatasetLink) saveAndReturnObject(ctx, link);
+            return new DatasetData(link.getChild());
+        } else {
+            return (DatasetData) saveAndReturnObject(ctx, dataset);
+        }
+    }
+    
+    /**
+     * Uploads and attaches a file to the provided {@link DataObject} (if
+     * provided)
+     * 
+     * @param ctx
+     *            {@link SecurityContext}
+     * @param file
+     *            The {@link File} to upload/attach
+     * @param mimetype
+     *            The mime type of the file (can be <code>null</code>)
+     * @param description
+     *            The description (can be <code>null</code>)
+     * @param namespace
+     *            The namespace (can be <code>null</code>)
+     * @param target
+     *            The {@link DataObject} to attach the file to (can be
+     *            <code>null</code>)
+     * @param callback
+     *            If no {@link Callback} is provided, the method will block
+     *            until the task is finished, otherwise the upload runs
+     *            asynchronously and the {@link Callback} handle gets notified
+     *            about the outcome
+     * @return The {@link FileAnnotationData} if no {@link Callback} handle was
+     *         provided, <code>null</code> otherwise
+     */
+    public FileAnnotationData attachFile(final SecurityContext ctx,
+            final File file, String mimetype, final String description,
+            final String namespace, final DataObject target, Callback callback) {
+        final String name = file.getName();
+        String absolutePath = file.getAbsolutePath();
+        final String path = absolutePath.substring(0, absolutePath.length()
+                - name.length());
+
+        final String mime;
+        if (mimetype == null)
+            mime = "application/octet-stream";
```

Comment:
Might be nice to use http://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#probeContentType%28java.nio.file.Path%29.